### PR TITLE
ZCS-11808: Display feedback when any action is in progress (new volume wizard)

### PR DESCRIPTION
**Pull request by luisenriquez01**
Display loading modal component to provide feedback whenever an operation is happening in the backend. This covers the following:

1. ValidateS3BucketReachableRequest
2. CreateS3BucketConfigRequest
3. GetS3BucketConfigRequest

While the modal is visible, the user cannot interact with the rest of the UI.

### CODE REVIEW

**Comment by luisenriquez01** [comment]
@k-kato should we worry about CSS compatibility here?

File: WebRoot/messages/ZaMsg.properties [hunk 1]
```diff
@@ -2332,6 +2332,9 @@ ss_inactive_accounts_30 = Inactive Accounts (30 days)
 ss_inactive_accounts_90 = Inactive Accounts (90 days)
 
 #Busy Messages
+BUSY_LOADING = Loading...
+BUSY_TESTING = Testing...
+BUSY_CREATING = Creating...
```
Comment (k-kato) [inline WebRoot/messages/ZaMsg.properties:2337]: BUSY_TESTING and BUSY_CREATING are not used.

File: WebRoot/templates/admin/Widgets.template [hunk 2]
```diff
@@ -35,6 +35,12 @@
 			</td>
 			<td id='${id}_settingimg'						style='width:20px'></td>
 			<td id='${id}_dropdown'	class='ZDropDown'		style='width:20px'></td>
+			<td id='loadingStatusBar' class='loadingContainer' style='visibility:hidden;'>
+				<div style='position:relative;height:inherit;'>
+					<div class='LoaderTitle'>Loading...</div>
```
Comment (k-kato) [inline WebRoot/templates/admin/Widgets.template:40]: please don't set non-translated string in a template.

File: WebRoot/templates/admin/Widgets.template [hunk 3]
```diff
@@ -82,12 +88,18 @@
...
-			<tr class='DwtDialogHeader' id='${dragId}'>
+			<tr class='DwtDialogHeader' id='${dragId}' style='position:relative;'>
 				<td class='minWidth'>${icon}</td>
 				<td id='${id}_title' class='DwtDialogTitle'>${title}</td>
 				<td class='minWidth'><div class='${closeIcon2}'></div></td>
 				<td class='minWidth'><div class='${closeIcon1}'></div></td>
 				<td width=24 id='${id}_minimize' class='DwtDialogMinIcon'></td>
+				<td id='loadingStatusBar' class='loadingContainer' style='visibility:hidden;'>
+					<div style='position:relative;height:inherit;'>
+						<div class='LoaderTitle'>Loading...</div>
```
Comment (k-kato) [inline WebRoot/templates/admin/Widgets.template:99]: please don't set non-translated string in a template.

File: WebRoot/js/zimbraAdmin/common/ZaApp.js [hunk 4]
```diff
@@ -95,6 +95,57 @@ ZaApp.checkMyRight = function(targetType,targetBy,targetVal,right,attrs) {
 	
 }
 
+ZaApp.toggleLoadingStatusBar = function (instance, show, label) {
+	var elementId = instance._htmlElId;
+	var element;
+	if (elementId) {
+		element = document.getElementById(elementId);
+		var localButtonTree = element.querySelectorAll(".xform_button");
```
Comment (k-kato) [inline WebRoot/js/zimbraAdmin/common/ZaApp.js:103]: Are only the button(s) of `xform_button` class the scope of the control? The function should be more generic if it is  defined in ZaApp. If it is difficult, target buttons or items should be enabled/disabled in a view/controller and this function should just show/hide the loading status bar.

**Comment by k-kato** [comment]
> @k-kato should we worry about CSS compatibility here?

Yes, it needs be tested on all supported browsers; at least MS Edge, MS  Edge with IE mode, Firefox, Safari and Google Chrome.
https://zimbra.github.io/zimbra-9/single-server-install.html#System_Requirements

File: WebRoot/js/zimbraAdmin/common/ZaApp.js [hunk 5]
```diff
@@ -95,6 +95,57 @@ ZaApp.checkMyRight = function(targetType,targetBy,targetVal,right,attrs) {
 	
 }
 
+ZaApp.toggleLoadingStatusBar = function (instance, show, label) {
```
[the remark this answers is not in the document]
Comment (luisenriquez01) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js:98]: Since it's used as a global function, it's easier to call ZaApp.toggleLoadingStatusBar than ZaApp.getInstance().toggleLoadingStatusBar. In this case, the instance parameter passed is that one of the current view (for this particular ticket, it's ZaNewVolumeXWizard).

File: WebRoot/messages/ZaMsg.properties (hunk 1 above)
Comment (luisenriquez01) [reply WebRoot/messages/ZaMsg.properties:2337]: They are used in this PR: https://github.com/Zimbra/zm-hsm-admin-zimlet/pull/29

File: WebRoot/templates/admin/Widgets.template (hunk 2 above)
Comment (luisenriquez01) [reply WebRoot/templates/admin/Widgets.template:40]: Could I leave this empty? It is getting replaced with the .properties value in the toggleLoadingStatusBar method.

File: WebRoot/templates/admin/Widgets.template (hunk 3 above)
Comment (luisenriquez01) [reply WebRoot/templates/admin/Widgets.template:99]: Same as above.

File: WebRoot/js/zimbraAdmin/common/ZaApp.js (hunk 4 above)
Comment (luisenriquez01) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js:103]: I'll review that with Saket later today. For this use case, .xform_button elements are all those set as type: _BUTTON_ within a form.

Comment (k-kato) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js:103]: I mean, is it okay not to disable other items like input field? Just need to disable buttons only?

File: WebRoot/js/zimbraAdmin/common/ZaApp.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (k-kato) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js:98]: okay

File: WebRoot/messages/ZaMsg.properties (hunk 1 above)
Comment (k-kato) [reply WebRoot/messages/ZaMsg.properties:2337]: It should be passed as `label` from a caller.
```
loaderTitle.innerHTML = label ? label : ZaMsg.BUSY_LOADING;
```
We can remove the two from here.

File: WebRoot/templates/admin/Widgets.template (hunk 2 above)
Comment (k-kato) [reply WebRoot/templates/admin/Widgets.template:40]: I think we can pass a string in ZaMsg like 
https://github.com/Zimbra/zm-admin-console/blob/develop/WebRoot/templates/dwt/Widgets.template#L530
(it's AjxMsg, though)

either is fine; empty value or ZaMsg.BUSY_LOADING.

File: WebRoot/messages/ZaMsg.properties (hunk 1 above)
Comment (k-kato) [reply WebRoot/messages/ZaMsg.properties:2337]: I reconsidered. That's fine. Please ignore my previous comment.

File: WebRoot/templates/admin/Widgets.template (hunk 2 above)
Comment (luisenriquez01) [reply WebRoot/templates/admin/Widgets.template]: Ready. Replaced with ZaMsg.BUSY_LOADING

File: WebRoot/js/zimbraAdmin/common/ZaApp.js (hunk 4 above)
Comment (luisenriquez01) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js]: @k-kato this was reviewed with @dawoodshaikh and @joshualandsman today and it was agreed that all buttons and input fields should be disabled. I have updated the code to cover all items regardless of their class.
Note: the `ZaMsg.BUSY_TESTING` has changed to `ZaMsg.BUSY_VALIDATING`. Loading and Creating are kept since they are going to be used in ZCS-11944.

Comment (k-kato) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js]: I checked the latest (= 2nd) commit and it looks good. 
One more question. Will a pull down menu button be disabled as well? For example, "Region" in "Add New Volume" dialog. I don't think it is a button or an input field...

If it is not easy to handle it and/or it is not important to disable such a menu, we can address it on a separate ticket in the near future.

Comment (luisenriquez01) [reply WebRoot/js/zimbraAdmin/common/ZaApp.js]: It wouldn't be simple to disable those fields programmatically, but a workaround would be to set CSS `pointer-events: none` so that they are not clickable.

File: WebRoot/js/zimbraAdmin/common/ZaApp.js [hunk 6]
```diff
@@ -95,6 +95,66 @@ ZaApp.checkMyRight = function(targetType,targetBy,targetVal,right,attrs) {
...
```
Comment (silentsakky) [inline WebRoot/js/zimbraAdmin/common/ZaApp.js:98]: instead of introducing a new UX pattern, can we use same pattern that we use for showing a popup when loading large data? by having different UX patterns we are loosing on consistency in product

**Review by silentsakky — APPROVED** [review]

**Review by lavegupta — APPROVED** [review]

**Review by k-kato — APPROVED** [review]
